phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/PagingRegionScanner.java

```diff
@@ -101,20 +149,28 @@ private boolean next(List<Cell> results, boolean raw) throws IOException {
                         LOGGER.info("Page filter stopped, generating dummy key {} ",
                                 Bytes.toStringBinary(rowKey));
                         ScanUtil.getDummyResult(rowKey, results);
+                        --multiKeyPointLookupPosition;
                     }
                     return true;
                 }
-                return false;
+                return isMultiKeyPointLookup ? multiKeyPointLookupPosition < multiKeyPointLookupRanges.size() : false;
             } else {
                 // We got a row from the HBase scanner within the configured time (i.e., the page size). We need to
                 // start a new page on the next next() call.
-                return true;
+                return isMultiKeyPointLookup ? multiKeyPointLookupPosition < multiKeyPointLookupRanges.size() : true;
             }
         } catch (Exception e) {
             if (pagingFilter != null) {
                 pagingFilter.init();
             }
+            if (isMultiKeyPointLookup) {
+                --multiKeyPointLookupPosition;
+            }
             throw e;
+        } finally {
+            if (isMultiKeyPointLookup) {
+                delegate.close();
+            }
```

Comment:
Can we close the delegate within `if (isMultiKeyPointLookup)` condition, just to maintain the pattern similar to adjusted scan boundaries?

Reply:
We also need to close always even an exception is thrown. Please note we close the current scanner after scanning a row. In adjusted scan boundaries, we close the previous scanner not the one to be scanned. So, closing it at finally makes sense. Does not it?

Reply:
Sounds good